File: acorn-core/src/test/java/com/ardley/acorn/policy/EvaluationPolicyTest.java
package com.ardley.acorn.policy;

import com.ardley.acorn.attribute.Attributes;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import java.util.Optional;

import static org.assertj.core.api.Assertions.assertThat;

/**
 * Tests isolation policy enforcement logic across various attribute configurations.
 */
class EvaluationPolicyTest {

    @Test
    @DisplayName("Matching isolation attribute passes")
    void matchingAttributePasses() {
        EvaluationPolicy policy = EvaluationPolicy.withIsolation("tenant_id");
        var principal = Attributes.builder().with("tenant_id", "t-1").build();
        var resource = Attributes.builder().with("tenant_id", "t-1").build();

        assertThat(policy.checkIsolation(principal, resource)).isEmpty();
    }

    @Test
    @DisplayName("Mismatched isolation attribute produces violation")
    void mismatchedAttributeViolates() {
        EvaluationPolicy policy = EvaluationPolicy.withIsolation("tenant_id");
        var principal = Attributes.builder().with("tenant_id", "t-1").build();
        var resource = Attributes.builder().with("tenant_id", "t-2").build();

        Optional<String> result = policy.checkIsolation(principal, resource);

        assertThat(result).isPresent();
        assertThat(result.get()).contains("Isolation violation").contains("tenant_id");
    }

    @Test
    @DisplayName("Resource lacking isolation attribute is treated as unscoped (passes)")
    void resourceWithoutAttributePasses() {
        EvaluationPolicy policy = EvaluationPolicy.withIsolation("tenant_id");
        var principal = Attributes.builder().with("tenant_id", "t-1").build();
        var resource = Attributes.empty();

        assertThat(policy.checkIsolation(principal, resource)).isEmpty();
    }

    @Test
    @DisplayName("Principal lacking isolation attribute when resource has it is a violation")
    void principalMissingAttributeViolates() {
        EvaluationPolicy policy = EvaluationPolicy.withIsolation("tenant_id");
        var principal = Attributes.empty();
        var resource = Attributes.builder().with("tenant_id", "t-1").build();

        Optional<String> result = policy.checkIsolation(principal, resource);

        assertThat(result).isPresent();
        assertThat(result.get()).contains("principal lacks it");
    }

    @Test
    @DisplayName("No isolation policy performs no checks regardless of attributes")
    void noIsolationPerformsNoChecks() {
        EvaluationPolicy policy = EvaluationPolicy.none();
        var principal = Attributes.builder().with("tenant_id", "t-1").build();
        var resource = Attributes.builder().with("tenant_id", "t-DIFFERENT").build();

        assertThat(policy.checkIsolation(principal, resource)).isEmpty();
    }

    @Test
    @DisplayName("Multiple isolation attributes: first violation short-circuits")
    void multipleIsolationAttributesFirstViolationFails() {
        EvaluationPolicy policy = EvaluationPolicy.builder()
                .withIsolation("org_id")
                .withIsolation("tenant_id")
                .build();
        var principal = Attributes.builder().with("org_id", "org-X").with("tenant_id", "t-1").build();
        var resource = Attributes.builder().with("org_id", "org-Y").with("tenant_id", "t-1").build();

        Optional<String> result = policy.checkIsolation(principal, resource);

        assertThat(result).isPresent();
        assertThat(result.get()).contains("org_id");
    }

    @Test
    @DisplayName("Multiple isolation attributes: all must match")
    void multipleIsolationAllMustMatch() {
        EvaluationPolicy policy = EvaluationPolicy.builder()
                .withIsolation("org_id")
                .withIsolation("tenant_id")
                .build();
        var principal = Attributes.builder().with("org_id", "org-1").with("tenant_id", "t-1").build();
        var resource = Attributes.builder().with("org_id", "org-1").with("tenant_id", "t-1").build();

        assertThat(policy.checkIsolation(principal, resource)).isEmpty();
    }
}

File: node/packages/core/src/__tests__/evaluation-policy.test.ts
import { describe, test, expect } from 'bun:test';
import { EvaluationPolicy } from '../evaluation-policy.js';
import { Attributes } from '../attributes.js';

describe('EvaluationPolicy', () => {
  test('none() passes any combination', () => {
    const policy = EvaluationPolicy.none();
    const principal = Attributes.from({ tenant_id: 'a' });
    const resource = Attributes.from({ tenant_id: 'b' });
    expect(policy.checkIsolation(principal, resource)).toBeUndefined();
  });

  test('withIsolation denies mismatched attribute', () => {
    const policy = EvaluationPolicy.withIsolation('tenant_id');
    const principal = Attributes.from({ tenant_id: 'a' });
    const resource = Attributes.from({ tenant_id: 'b' });
    const result = policy.checkIsolation(principal, resource);
    expect(result).toContain('Isolation violation');
    expect(result).toContain("principal='a'");
    expect(result).toContain("resource='b'");
  });

  test('passes when resource lacks isolation attribute', () => {
    const policy = EvaluationPolicy.withIsolation('tenant_id');
    const principal = Attributes.from({ tenant_id: 'a' });
    const resource = Attributes.from({ other: 'x' });
    expect(policy.checkIsolation(principal, resource)).toBeUndefined();
  });

  test('denies when resource has attr but principal lacks it', () => {
    const policy = EvaluationPolicy.withIsolation('tenant_id');
    const principal = Attributes.from({});
    const resource = Attributes.from({ tenant_id: 'b' });
    const result = policy.checkIsolation(principal, resource);
    expect(result).toContain('principal lacks it');
  });

  test('passes when both match', () => {
    const policy = EvaluationPolicy.withIsolation('tenant_id');
    const principal = Attributes.from({ tenant_id: 'same' });
    const resource = Attributes.from({ tenant_id: 'same' });
    expect(policy.checkIsolation(principal, resource)).toBeUndefined();
  });

  test('builder pattern', () => {
    const policy = EvaluationPolicy.builder()
      .withIsolation('tenant_id')
      .withIsolation('org_id')
      .build();
    expect(policy.getIsolationAttributes()).toEqual(['tenant_id', 'org_id']);
  });

  test('strict mode denies when resource lacks isolation attribute', () => {
    const policy = EvaluationPolicy.withIsolation('tenant_id').strict();
    const principal = Attributes.from({ tenant_id: 'a' });
    const resource = Attributes.from({ other: 'x' });
    const result = policy.checkIsolation(principal, resource);
    expect(result).toContain('strict mode');
    expect(result).toContain("'tenant_id'");
  });

  test('strict mode still passes when both match', () => {
    const policy = EvaluationPolicy.withStrictIsolation('tenant_id');
    const principal = Attributes.from({ tenant_id: 'same' });
    const resource = Attributes.from({ tenant_id: 'same' });
    expect(policy.checkIsolation(principal, resource)).toBeUndefined();
  });

  test('strict mode denies mismatched attribute (unchanged from lenient)', () => {
    const policy = EvaluationPolicy.withStrictIsolation('tenant_id');
    const principal = Attributes.from({ tenant_id: 'a' });
    const resource = Attributes.from({ tenant_id: 'b' });
    const result = policy.checkIsolation(principal, resource);
    expect(result).toContain('Isolation violation');
    expect(result).toContain("principal='a'");
  });

  test('strict mode reports the first missing attribute across a multi-attribute policy', () => {
    const policy = EvaluationPolicy.builder()
      .withIsolation('tenant_id')
      .withIsolation('org_id')
      .strict()
      .build();
    const principal = Attributes.from({ tenant_id: 'a', org_id: 'o' });
    const resource = Attributes.from({ tenant_id: 'a' });
    const result = policy.checkIsolation(principal, resource);
    expect(result).toContain('strict mode');
    expect(result).toContain("'org_id'");
  });

  test('isStrict reflects mode', () => {
    expect(EvaluationPolicy.withIsolation('tenant_id').isStrict()).toBe(false);
    expect(EvaluationPolicy.withStrictIsolation('tenant_id').isStrict()).toBe(true);
    expect(EvaluationPolicy.builder().withIsolation('tenant_id').strict().build().isStrict()).toBe(true);
  });
});
